File: rnamake/secondary_structure_tree.py
import tree

def tree_from_pose(p, start_m=None):
    sst = SecondaryStructureTree()
    for m in p.motifs:
        found = 0
        for m1 in p.motifs:
            if m1 == m:
                continue
            for e in m1.ends:
                if e == m.ends[0]:
                    found = 1
                    break
        if not found:
            start_m = m
            break

    if start_m is None:
        raise ValueError("cannot convert pose to tree there are no motifs with free ends")
    open_motifs = [[start_m, -1, -1]]
    seen_motifs = {}
    while len(open_motifs) > 0:
        current, parent_index, parent_end_index = open_motifs.pop(0)
        seen_motifs[current] = 1

        pos = sst.add_motif(current, parent_index, parent_end_index)
        for m in p.motifs:
            if m in seen_motifs:
                continue
            for i, end1 in enumerate(current.ends):
                if end1 == 0:
                    continue
                if end1 == m.ends[0]:
                    open_motifs.append([m, pos, i])
    return sst


class SecondaryStructureTree(object):
    def __init__(self):
        self.tree = tree.TreeStatic()

    def add_motif(self, m, parent_index=-1, parent_end_index=-1, m_end_index=0,
                  parent_end_name=None):
        parent = self.tree.last_node
        if parent_index != -1:
            parent = self.tree.get_node(parent_index)

        m_copy = m.copy()
        if parent is None:
            return self.tree.add_data(m_copy, len(m_copy.ends), -1, -1)

        avail_pos = self.tree.get_available_pos(parent, parent_end_index)
        for p in avail_pos:
            pos = self.tree.add_data(m_copy, len(m_copy.ends),
                                     parent.index, p)
            return pos

        return -1

File: rnamake/lib/RNAMake/src/secondary_structure/secondary_structure_tree.cpp
//
//  secondary_structure_tree.cpp
//  RNAMake
//
//  Created by Joseph Yesselman on 12/14/15.
//  Copyright (c) 2015 Joseph Yesselman. All rights reserved.
//


#include <queue>

//RNAMake Headers
#include "secondary_structure/secondary_structure_tree.h"


namespace secondary_structure {
    
int
SecondaryStructureTree::add_motif(
    MotifOP const & m,
    int parent_index,
    int parent_end_index) {
    
    auto parent = tree_.last_node();
    
    if(parent_index != -1) {
        parent = tree_.get_node(parent_index);
    }
    
    auto m_copy = std::make_shared<Motif>(*m);
    if(parent == nullptr) {
        return tree_.add_data(m_copy, (int)m_copy->ends().size(), -1, -1);
    }
    
    auto avail_pos = tree_.get_available_pos(parent, parent_end_index);
    if(avail_pos.size() == 0) {
        throw std::runtime_error("cannot add motif to tree no available pos");
    }
    return tree_.add_data(m_copy, (int)m_copy->ends().size(), parent->index(), avail_pos[0]);
    
    return -1;
    
}
    
struct TreePreNode {
    MotifOP m;
    int parent_index, parent_end_index;
    
    inline
    TreePreNode() {}
    
    inline
    TreePreNode(
        MotifOP const & nm,
        int nparent_index,
        int nparent_end_index):
    m(nm),
    parent_index(nparent_index),
    parent_end_index(nparent_end_index)
    {}
};

SecondaryStructureTreeOP
tree_from_pose(PoseOP const & p) {
    auto sst = std::make_shared<SecondaryStructureTree>();
    auto start_m = MotifOP(nullptr);
    auto found = 0;
    for(auto const & m : p->motifs()) {
        found = 0;
        for(auto const & m1 : p->motifs()) {
            if(m == m1) { continue; }
            for(auto const & e : m1->ends()) {
                if(e == m->ends()[0]) {
                    found = 1;
                    break;
                }
            }
        }
        if(found == 0) {
            start_m = m;
            break;
        }
    }
    
    if(start_m == nullptr) {
        std::runtime_error("could not find start motif to build tree in secondary_structure::tree_from_pose");
    }
    
    auto open_nodes = std::queue<TreePreNode>();
    auto seen = MotifOPs();
    auto current = TreePreNode();
    open_nodes.push(TreePreNode{ start_m, -1 , -1});
    int i = 0;
    
    while(! open_nodes.empty()) {
        current = open_nodes.front();
        open_nodes.pop();
        
        seen.push_back(current.m);
        
        int pos = sst->add_motif(current.m, current.parent_index, current.parent_end_index);
        for(auto const & m : p->motifs()) {
            if(std::find(seen.begin(), seen.end(), m) != seen.end()) {
                continue;
            }
            i = -1;
            for(auto const & e : current.m->ends()){
                i++;
                if(i == 0) { continue; }
                if(e == m->ends()[0]) {
                    open_nodes.push(TreePreNode(m, pos, i));
                }
            }
        }
    }
    
    return sst;
}
    
    
}





















